core/templates/pages/release-coordinator-page/jobs-tab/jobs-tab.component.spec.ts

```diff
@@ -13,20 +13,19 @@
 // limitations under the License.
 
 /**
- * @fileoverview Unit tests for AdminJobsTabComponent
+ * @fileoverview Unit tests for JobsTabComponent
  */
 
 import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
 import { MatCardModule } from '@angular/material/card';
 
-import { AdminBackendApiService, AdminPageData } from 'domain/admin/admin-backend-api.service';
+import { ReleaseCoordinatorBackendApiService, JobsData } from '../services/release-coordinator-backend-api.service';
```

Comment:
Use absolute path.

Reply:
We are allowed to use relative path for internal services/components, I see this pattern in all the places.

@srijanreddy98 Can you confirm this?

Reply:
As a general rule, it is better to always use absolute paths. It avoids issues when we move files from one dir to another.

If we do this inconsistently I would suggest filing an issue to fix it in general. Could either you or @srijanreddy98 please make sure this happens?

Reply:
> As a general rule, it is better to always use absolute paths. It avoids issues when we move files from one dir to another.
Technically no solution is immune to this problem. When moving a folder around, absolute paths break too. There is a better to this. We should just look at import statements as something you are importing but also as a way to identify whether or not an import belongs to the unit this file is a part of.

This might be up for debate:
# Absolute vs Relative
Lets look at the best case examples for both these strategies:
## Absolute
imports from node_modules. This will always be absolute as node_modules don't belong to our application at all and in no case of development can we consider them to be a part of the unit (except after the final build, where imports don't matter.

## Relative:
Let assume we have a service called MockService. file: mock.service.ts. spec: mock.service.spec.ts. When importing MockService class inside the spec should be relative because no matter what, the service will always be in `./ ` of its spec file. Absolutes imports here would break when we move this service to a different folder while relative import in this case is immune.

These are the best cases and have a clear path. Others need more reasoning and clearer thought.

I am going draw the current scenario and then the possible file/folder move options with the results of each strategy:

## Current Scenario (might also look like my assumptions):
We have this jobs component, as a part of release-coordinator-page. No other page should use it. The services folder inside the release-coordinator-page folder, contain services that are only meant to be used inside the release-coordinator-page.

## Folder/file move 1: This component now has to be used in admin page too.

The best course of action, move this component to shared components.

### Case 1: Absolute imports:
No errors for the ReleaseCoordinatorBackendApiService it was imported using an absolute path.

### Case2: Relative imports:
We get errors saying that there is no module/file `../services/release-coordinator-backend-api.service`.

### Expected behaviour:
(Debatable) I would have definitely wanted an error because pages are built on top of shared components and not vice versa. When we use relative imports, it tells us that the particular files are "internal" to this folder and not "exposed"/"public". Each page should be independent of each other and if anything is shared, it should be in shared components. By this design, ReleaseCoordinatorBackendApiService should not be a part of jobs component if they both reside in different folder. We should have clear boundaries regarding this.

## Folder/file move 2: This component now has to be a part of a different page and is no-longer required here.

### Case 1: Absolute imports:
No errors for the ReleaseCoordinatorBackendApiService it was imported using an absolute path.

### Case2: Relative imports:
We get errors saying that there is no module/file `../services/release-coordinator-backend-api.service`.

Expected: (Debatable) Same as the previous case.

## Folder/file move 3: We rename the page folder:

### Case 1: Absolute imports:
We get errors because the absolute path doesn't exist.

### Case2: Relative imports:
No errors.

Expected: No errors

## Conclusion

From these three examples it might look like relative is better but its not. There are many more cases where absolute imports easily trump over relatives. Its not evident here because the better "import strategy" here for this particular case is relative.

I would like to sum it all up with a rule that uses the best of both worlds:

**This rule has a prerequisite that needs to properly define the word "unit" and an ill defined unit will be detrimental to the best outcome.**

## Optimal import strategy (IMO) 

1> Relative imports for cases where the import belongs to the same unit.
2> Absolute import everywhere else.

My suggestion for units in our codebase:
1> Every first level folder inside pages folder.
2> Every first level folder inside shared-components folder along with the services folder.
3> Every subfolder in the extensions/interactions folder
4> The extensions/object-editors folder
5> The extensions/rte folder (don't know the exact path but hopefully we'll get the gist).

A general pattern that can be found in the "definition of unit is that, we have a module.ts file in those folders define a "module" or according to the definition we want to set, unit.

If you all have other scenarios related to this particular import, I am happy to discuss the consequences of having either import strategy and also the one that I just proposed.

Reply:
Thanks @srijanreddy98. Let me defer to you and @DubeySandeep to come up with a general rule that is easy for contributors to follow and doesn't have a lot of special cases. We'll then go with that, but please make sure it's documented clearly, broadcast as needed, and that contributors can easily follow it.

The only things I'd point out that might be worth adding to the data points for consideration are the following:

1. Absolute imports are easier to verify at a glance, whereas relative ones require the reader to have some additional contextual knowledge of the base path. The latter might be OK if "relative" is always relative to the location of the current file (and not to some dir that is somehow dependent on the execution context).
2. A definition of "units" that has 5 different cases won't do -- too much to keep track of. If a unit is simply defined as the root dir containing a module.ts file then this might work since it's a single criterion that's easy enough to remember. But this definition must always work and end up being a rule that we follow (possibly enforced with lint checks) -- it can't work in 90% of the cases with some exception for the others.

Incidentally, are there any general Angular best practices for this? There is some argument to be made for following what "standard practice" is, but I don't know if things have been solidified to this level yet in the frontend world.

Reply:
> Incidentally, are there any general Angular best practices for this? There is some argument to be made for following what "standard practice" is, but I don't know if things have been solidified to this level yet in the frontend world.

Not that I have seen. This is something that I came up with for our codebase. But they do extensively use relative imports if the imports are from the same "package". https://github.com/angular/angular/blob/master/packages/core/src/render3/jit/directive.ts
The relative imports do go beyond core/src. All of the imports are from inside core/src.

Reply:
> Let me defer to you and @DubeySandeep to come up with a general rule that is easy for contributors to follow and doesn't have a lot of special cases.

@seanlip I'm working with @srijanreddy98 to document the rule and add a lint check for the same.

Reply:
@seanlip, I've created an issue for the requirements #12966

@srijanreddy98 I've assigned you to the issue, I hope that's fine. Let me know if I can help anyway! :)

Reply:
Yup, no problem!